### DISCUSSION TITLE
prometheus Monitoring

### DISCUSSION THREAD

**Discussion by Juandewet** [netinvent/npbackup#69] [Q&A]
Hey I'm having some difficulty understating how exactly Prometheus gets the metrics form endpoints, I'm very new to this and was hoping somebody could help me.

I have a REST server, Prometheus and Grafana docker stack on the same network

**Comment by deajan** [comment]
Hi @Juandewet, sorry, I'm quite offline for a couple of days, so sorry for the response delay.

Basically, NPBackup will create prometheus metrics in two forms, depending on what you configure the metrics target:
- Path to file: Write a metrics file that will be picked up by node_exporter / windows_exporter
- URL: Upload metrics to a prometheus push gateway, which can be interrogated by Prometheus later

For the latter, you need to setup a [prometheus push gateway](https://github.com/prometheus/pushgateway), with preferably authentication and https support. I use Haproxy to setup auth and https.

I can be of assistance for more documentation, but not before next week ;)

**Comment by GuitarBilly** [comment]
hello @deajan,
can you give some examples for the pushgateway URL in the npbackup config window? 
I cannot get that to work, metrics do not end up in the pushgateway...

p.s. node_exporter works for me and npbackup is a much more useful backup status dashboard than the example provided at the rest-server github page (which is basically a performance dashboard). 
thanks and well done!

**Reply by deajan** [reply]
Thank you for your kind words.
Basically, the URL I'm using is something like `https://push.mydomain.tld/metrics/job/${BACKUP_JOB}` where push.mydomain.tld is the address of the push gateway.
I'll have this documented.
Let met know if you need further help.

**Reply by GuitarBilly** [reply]
@deajan,
perfect, now it works! It clearly shows my lack of experience with prometheus.

( Next step for me is to upgrade to the latest npbackup v3 development and try that dashboard too. )

FYI, I now have a setup at home that consists of
- server machine A with linux and rest-server, prometheus, node_exporter, pushgateway and grafana via docker compose.
- workstation machine B with linux and npbackup.
- laptop machine C, D, E with W10 and restic via the windows scheduler.

To my understanding i'd use the node_exporter for my server A. And use the pushgateway for backups started from Workstation B and laptops C,D,E (which I will upgrade to npbackup for metrics and browse/restore gui bonus).
Is that how you also deploy it?

**Reply by deajan** [reply]
Yes, that's mostly the way I do.
I've made a basic monitoring doc on the github wiki. Feel free to comment if that works for you.

**Reply by GuitarBilly** [reply]
thank you. 
the wiki is a nice summary to guide users. hopefully TS is also helped with it!
